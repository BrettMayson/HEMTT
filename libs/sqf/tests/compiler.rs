#![allow(clippy::unwrap_used)]

pub use float_ord::FloatOrd as Scalar;
use hemtt_preprocessor::Processor;
use hemtt_sqf::{compiler::serializer::Compiled, parser::database::Database};
use hemtt_workspace::LayerType;

macro_rules! compile {
    ($dir: expr, $file:ident) => {
        paste::paste! {
            #[test]
            fn [<simple_ $file>]() {
                let (bin, compiled) = optimize($dir, stringify!($file));
                // bin to hex
                let hex = bin.iter().map(|b| format!("{:02x}", b)).collect::<String>();
                insta::assert_snapshot!(hex);
                insta::assert_debug_snapshot!(compiled);
            }
        }
    };
}

compile!("optimizer", consume_array);
compile!("optimizer", static_math);
compile!("optimizer", scalar);
compile!("optimizer", string_case);
compile!("simple", dev);
compile!("simple", eventhandler);
compile!("simple", foreach);
compile!("simple", format_font);
compile!("simple", get_visibility);
compile!("simple", hash_select);
compile!("simple", hello);
compile!("simple", include);
compile!("simple", oneline);
compile!("simple", semicolons);

const ROOT: &str = "tests/";

fn optimize(folder: &str, file: &str) -> (Vec<u8>, Compiled) {
    let folder = std::path::PathBuf::from(ROOT).join(folder);
    let workspace = hemtt_workspace::Workspace::builder()
        .physical(&folder, LayerType::Source)
        .finish(None, false, &hemtt_common::config::PDriveOption::Disallow)
        .unwrap();
    let source = workspace.join(format!("{file}.sqf")).unwrap();
    let processed = Processor::run(&source).unwrap();
    let mut writer = Vec::new();
    let statements = hemtt_sqf::parser::run(&Database::a3(false), &processed).unwrap();
    let compiled = statements.compile(&processed).unwrap();
    statements
        .optimize()
        .compile_to_writer(&processed, &mut writer)
        .unwrap();
    (writer, compiled)
}
