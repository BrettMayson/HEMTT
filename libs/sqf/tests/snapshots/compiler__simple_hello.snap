---
source: libs/sqf/tests/compiler.rs
expression: hex
---
010000000520000000023103000500005f6e616d65060000666f726d61740a000073797374656d6368617411000001bb00000002001604000105000048454d5454010a000048656c6c6f2c2025312101570000236c696e65203120225c687c0200052e737166220a70726976617465205f6e616d65203d202290070005223b0a73797374656d4368617420666f726d6174205b22283e01222ca406035d3b0a000200a100097800020100000615c001040000010100074f74016e040841a4010004020002390000000002000100022e00000000020002001100000201000a00002f68656c6c6f2e737166030300000000000000
